### ISSUE TITLE
Question about Multi-GPU inference (greedy generation with stopping criteria)

### ISSUE THREAD

**Issue by sfc-gh-reyazda** [huggingface/transformers#30150]
Hi,

I am curious about the implementation of the gpu-syncing on the generation pipeline when using multiple GPUs. I have been reading the code for the `_greedy_search` function and noticed at one part of the generation loop, we are checking for the each device (peer) to see if it is finished with the generation (potentially reaching a stopping condition) and then it skips the new-token generation: https://github.com/huggingface/transformers/blob/main/src/transformers/generation/utils.py#L2444. However, I am wondering why we need to even run the forward of the model (https://github.com/huggingface/transformers/blob/main/src/transformers/generation/utils.py#L2436) if we are not using the output logits anymore.
Do you think it would have been better that we moved this condition at the beginning of the generation loop before preparing the input and running the forward? 

Also, another issue that I can think of by delaying the skipping of the generation iteration for this peer that is finished is that there is a chance that the rest of the devices finish the post-processing and therefore `_has_unfinished_sequences` returns `True`, in which case we run into a deadlock as the already finished peers are running the forward still and the system hangs for any communication that is happening in the model!

Any feedback is appreciated :)

Thanks,
Reza

**Comment by amyeroberts** [comment]
cc @gante

**Comment by gante** [comment]
Hey @sfc-gh-reyazda 👋 

The answer is here: https://huggingface.co/docs/transformers/v4.39.3/en/deepspeed#generate 🤗
